src/engine/sparqlExpressions/SparqlExpressionGenerators.h

```diff
@@ -124,27 +167,59 @@ inline auto valueGetterGenerator =
   return makeGenerator(AD_FWD(input), numElements, context, transformation);
 };
 
+template <typename FunctionType, typename... Generators>
+struct ApplyFunctionState {
+  using ResultType = std::decay_t<std::invoke_result_t<
+      FunctionType, ql::ranges::range_value_t<Generators>...>>;
+
+  std::decay_t<FunctionType> function_;
+  std::tuple<std::decay_t<Generators>...> generators_;
+  std::tuple<ql::ranges::iterator_t<std::decay_t<Generators>>...> iterators_;
+  size_t remaining_;
+
+  ApplyFunctionState(FunctionType f, size_t n, Generators... gens)
+      : function_(std::move(f)),
+        generators_(std::move(gens)...),
+        remaining_(n) {
+    iterators_ = std::apply(
+        [](auto&... gens) { return std::tuple{gens.begin()...}; }, generators_);
+  }
+};
+
+template <typename StateType>
+auto createValueProducer(std::shared_ptr<StateType> state) {
+  using ResultType = typename StateType::ResultType;
+  return [state]() mutable -> std::optional<ResultType> {
+    if (state->remaining_ == 0) {
+      return std::nullopt;
+    }
+    auto functionOnIterators = [&](auto&&... its) {
+      return state->function_(AD_MOVE(*its)...);
+    };
+    ResultType result = std::apply(functionOnIterators, state->iterators_);
+    std::apply([](auto&&... its) { (..., ++its); }, state->iterators_);
+    --state->remaining_;
+    return result;
+  };
+}
+
 /// Do the following `numItems` times: Obtain the next elements e_1, ..., e_n
 /// from the `generators` and yield `function(e_1, ..., e_n)`, also as a
 /// generator.
 inline auto applyFunction = [](auto&& function, size_t numItems,
-                               auto... generators)
-    -> cppcoro::generator<std::invoke_result_t<
-        decltype(function),
-        ql::ranges::range_value_t<decltype(generators)>...>> {
-  // A tuple holding one iterator to each of the generators.
-  std::tuple iterators{generators.begin()...};
-
-  auto functionOnIterators = [&function](auto&&... iterators) {
-    return function(AD_MOVE(*iterators)...);
-  };
```

Comment:
This is basically `views::zip(OwningView{std::move(generators), function})` , but I have to make sure that we can combine the `range-v3` with std::ranges.